supabase/async_client.py

```diff
@@ -0,0 +1,186 @@
+from .lib.client_options import ClientOptions
+from .lib.storage_client import SupabaseStorageClient
+
+
+class AsyncSupabaseClient:
+    """Supabase client class."""
+
+    def __init__(
+        self,
+        supabase_url: str,
+        supabase_key: str,
+        options: ClientOptions = ClientOptions(),
+    ):
+        """Instantiate the client.
+
+        Parameters
+        ----------
+        supabase_url: str
+            The URL to the Supabase instance that should be connected to.
+        supabase_key: str
+            The API key to the Supabase instance that should be connected to.
+        **options
+            Any extra settings to be optionally specified - also see the
+            `DEFAULT_OPTIONS` dict.
+        """
+
+        if not supabase_url:
+            raise SupabaseException("supabase_url is required")
+        if not supabase_key:
+            raise SupabaseException("supabase_key is required")
+
+        # Check if the url and key are valid
+        if not re.match(r"^(https?)://.+", supabase_url):
+            raise SupabaseException("Invalid URL")
+
+        # Check if the key is a valid JWT
+        if not re.match(
+            r"^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$", supabase_key
+        ):
+            raise SupabaseException("Invalid API key")
+
+        self.supabase_url = supabase_url
+        self.supabase_key = supabase_key
+
+        options.headers.update(self._get_auth_headers())
+
+        self.rest_url: str = f"{supabase_url}/rest/v1"
+        self.realtime_url: str = f"{supabase_url}/realtime/v1".replace("http", "ws")
+        self.auth_url: str = f"{supabase_url}/auth/v1"
+        self.storage_url = f"{supabase_url}/storage/v1"
+
+        is_platform = re.search(r"(supabase\.co)|(supabase\.in)", supabase_url)
+        if is_platform:
+            url_parts = supabase_url.split(".")
+            self.functions_url = (
+                f"{url_parts[0]}.functions.{url_parts[1]}.{url_parts[2]}"
+            )
+        else:
+            self.functions_url = f"{supabase_url}/functions/v1"
+
+        self.schema: str = options.schema
+
+        # Instantiate clients.
+        self.auth = self._init_supabase_auth_client(
+            auth_url=self.auth_url,
+            client_options=options,
+        )
+        # TODO: Bring up to parity with JS client.
+        # self.realtime: SupabaseRealtimeClient = self._init_realtime_client(
+        #     realtime_url=self.realtime_url,
+        #     supabase_key=self.supabase_key,
+        # )
+        self.realtime = None
+        self.postgrest = self._init_postgrest_client(
+            rest_url=self.rest_url,
+            supabase_key=self.supabase_key,
+            headers=options.headers,
+            schema=options.schema,
+        )
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.postgrest.session.aclose()
```

Comment:
1) We've to do all cleanup here, including closing the other clients as well (ex: `storage`)
2) You can use the [aclose](https://postgrest-py.readthedocs.io/en/latest/api/client.html#postgrest.AsyncPostgrestClient.aclose) method on the postgrest client here
```suggestion
        await self.postgrest.aclose()
```